gcc/clang on linux emit warnings

Hello,

first let me say that this library looks really awesome. Thanks for the great work.

I have nevertheless a slight problem when compiling with `gcc` or `clang` (with `-Wall`, `-Wextra`, and `-pedantic`). There are a very large amount of warnings (unknown-pragmas only with gcc, unused parameters, defined but not used variables, unused variables, etc).

I tested with gcc 11.4 and clang 14.0 on Ubuntu.

Unfortunately this is problematic for use for using in a testing framwork since it pollutes a lot the possible warnings of the actual library one wants to test.

Most problems are related with using linux I think and could be corrected with **a lot** of pre-processing directives spread all over the place.

What would you think about the idea of creating a MR about this?

Hi @fdefelici,

thanks a lot for your work!

I can confirm that the code in the branch <https://github.com/fdefelici/clove-unit/tree/issue-73-gcc-clang_on_linux_emit_warnings> (I think there is a typo in your comment above and I thought there was no modification at first :)) emits almost no warnings. To be more precise:

1. With gcc 13.2.1 no warnings (-Wall -Wextra -pedantic)
2. With clang 16.0.6 (not clang++, I'm more a "C" guy ;)) only one warning:

`./clove-unit.h:45:48: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
__CLOVE_EXTERN_C void __clove_utils_empty_funct();`

There are a few of them. But I find them weird, because there is no obvious reason for this to happen...

--- a/clove-unit.h
+++ b/clove-unit.h
@@ -13,11 +13,11 @@
 #define __CLOVE_VERSION_PATCH 1
 #define __CLOVE_VERSION "2.4.1"
 
-//Preventing "unknown-pragmas" warning on GCC for '#pragma region' usage
-//NOTE1: It seems that GCC v13+ should support '#pragma region' by the way.
-//NOTE2: This workaround doesn't work for G++ version < 13 because of a bug 
+//Preventing "unknown-pragmas" warning on GCC <= 12 for '#pragma region' usage
+//NOTE1: GCC and G++ v13+ support '#pragma region' by the way.
+//NOTE2: This workaround doesn't work for G++ version <= 12 because of a bug 
 //       that has been addressed only in G++ v13 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53431)
-#ifdef __GNUC__ // GNU Unknown Pragma Fix: START
+#if defined(__GNUC__) && __GNUC__ <= 12 // GNU Unknown Pragma Fix: START
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wunknown-pragmas"
 #endif 
@@ -3964,7 +3964,7 @@ void __clove_exec_suite(__clove_suite_t* suite, size_t test_counter, size_t* pas
 
 #pragma endregion //PUBLIC
 
-#ifdef __GNUC__ // GNU Unknown Pragma Fix: END
+#if defined(__GNUC__) && __GNUC__ <= 12 // GNU Unknown Pragma Fix: END
 #pragma GCC diagnostic pop
 #endif 
 

